### ISSUE TITLE
Feature request: download speed in logs

### ISSUE THREAD

**Issue by M1hacka** [pgbackrest/pgbackrest#429]
Hello.
I have s3 integration build with postgresql 9.4. Everything went fine for a week, but yesterday my standby server retention began growing. I can't understand why:
1. It's not wal push problem, all wals are on amazon now, `pg_xlog` is not growing and there are less `.ready` files in `pg_xlog/archive_status`
2. No fails in standby logs. It downloads WALs - one in 10 seconds, but it's too slow for our database.
3. The second thing we checked - disk read/write and internet connection problems. Everything is fine, according to our tests.
4. Amazon download issue? I've tried manual downloading wal via aws-cli, and it went fast. In order to investigate such problems, it would be good to output WAL download speed (or time) in log files, as [WAL-E](https://github.com/wal-e/wal-e), for example, does.
5. `pgbackrest check` works fine on all servers

I'll also be greatful for any suggestions on my standby problem.

**UPD:**
And also, it would be good to log whether WAL was downloaded from s3 or from primary server

**UPD 2**
Found `terminating walsender process due to replication timeout` in replica and walreceiver termination line in logs, at the moment when replication stopped working. Current theory is:
1. Standby got wals directly from Master
2. Some error occured, walsender and walreceiver processes failed on timeout. During this period some WALs were collected and Standby got retention from Master
3. When walsender and walreceiver rebooted, Standby began downloading WALs from Amazon and for some reason it works too slow, so retention grows up.

The questions are now: 
- How can I understand the reason of slow download from S3 (As I've written above, no speed problems with aws-cli were found)? 
- How can I speed up download (multiple threads, for instance?)

**Comment by M1hacka** [comment]
And also, it would be good to log whether WAL was downloaded from s3 or from primary server

**Comment by dwsteele** [comment]
It's not just about download speed, unfortunately.  pgBackRest does a number of cross-checks and has to load control files which takes time.  Download speed on AWS is fine, but request time can be miserable.

We are working on parallel async archive-get to compliment parallel async archive-push.  This will allow us to do many of the operations once for a larger number of WAL downloads and speed the process up a lot.  We don't have an ETC for that feature, though.

Meanwhile, if you have enough WAL on your master to bring the replica up to date, you can comment out `recover_command` in `recovery.conf` and this will force the standby to read from the master instead of the archive to catch up.

**Comment by M1hacka** [comment]
> pgBackRest does a number of cross-checks and has to load control files which takes time.

But what is the "bottleneck"? I don't see any kind of load on the server: neigther CPU, Disk IO or Network superactivity. 
1. Standby is cloud server, so if it is computational load, I could just add server resources. But there is no load at all. 
2. It would be strange, if it's network as master/standby/backup are in local network with speed over 500 Mbps, And network activity metrics show 50Mbps at max during WAL downloading...

**Comment by dwsteele** [comment]
After a bit of investigation I found the bottleneck.  Apparently, S3 rate limits list requests after 72K.  Here's a trace of listing a full WAL subdir with 256 files:
```
2017-10-19 20:13:36.210 P00   WARN: CONTENT LENGTH: -1
2017-10-19 20:13:36.226 P00   WARN: CHUNK LENGTH: 8192
2017-10-19 20:13:36.226 P00   WARN: CHUNK LENGTH: 8192
2017-10-19 20:13:36.226 P00   WARN: CHUNK LENGTH: 8192
2017-10-19 20:13:36.239 P00   WARN: CHUNK LENGTH: 8192
2017-10-19 20:13:36.239 P00   WARN: CHUNK LENGTH: 8192
2017-10-19 20:13:36.246 P00   WARN: CHUNK LENGTH: 8192
2017-10-19 20:13:36.246 P00   WARN: CHUNK LENGTH: 8192
2017-10-19 20:13:36.256 P00   WARN: CHUNK LENGTH: 8192
2017-10-19 20:13:36.256 P00   WARN: CHUNK LENGTH: 8192
<...HUGE PAUSE HERE...>
2017-10-19 20:13:42.259 P00   WARN: CHUNK LENGTH: 8192
2017-10-19 20:13:42.259 P00   WARN: CHUNK LENGTH: 2330
2017-10-19 20:13:42.259 P00   WARN: CHUNK LENGTH: 0
2017-10-19 20:13:42.259 P00   WARN: RESPONSE BODY LENGTH: 84250
```
The filename of the WAL file is not known when getting it (because of the appended hash) so the subdir is listed and then filtered with a regex (which s3 does not support).

This can be optimized by using any constant part of the regex after the ^ as a prefix to the list request, which will definitely help in this case at least.

**Comment by dwsteele** [comment]
So yeah, that seems better:
```
2017-10-19 20:46:41.748 P00   WARN: PREFIX IS 0000000100000129000000FF
2017-10-19 20:46:41.854 P00   WARN: CONTENT LENGTH: -1
2017-10-19 20:46:41.854 P00   WARN: CHUNK LENGTH: 331
2017-10-19 20:46:41.854 P00   WARN: CHUNK LENGTH: 0
2017-10-19 20:46:41.854 P00   WARN: RESPONSE BODY LENGTH: 331
```

All our automated testing is done with Scality S3 which does not do any rate limiting.  Automated testing is great for correctness but often does not reflect real world conditions...

**Comment by dwsteele** [comment]
The patch looks good.  How would you like your name to appear in the release notes?

**Comment by M1hacka** [comment]
Doesn't matter. I'm just looking forward to try the release and hope it will not take too long...

**Comment by dwsteele** [comment]
Well, I like to give credit where it is due.  Committed at 1f120f3fce1f8f30.

A release should be coming out next week, though I'm not 100% sure about that.

**Comment by dwsteele** [comment]
Did v1.25 help with performance?

**Comment by M1hacka** [comment]
Yesterday I switched to 1.25, today standby  retention began growing again...
According to logs, restore speed hasn't grown significantly (5-10 seconds for single log file)

**Comment by dwsteele** [comment]
Here's an example from one of my systems getting a 3.7mb (compressed) WAL file:
```
$ pgbackrest --stanza=XXX --log-level-console=debug archive-get 000000010000000B00000054 ~/TESTXLOG

2017-10-31 11:59:43.426 P00  DEBUG:     Storage::Local->openRead(): bIgnoreMissing = <false>, rhyFilter = [undef], xFileExp = <REPO:ARCHIVE>/9.4-1/000000010000000B00000054-b05a8fe51573d42b49e6bad71fd8ef44d05a9ee4.gz
2017-10-31 11:59:43.626 P00  DEBUG:     Storage::Posix::Driver->new(): bFileSync = <true>, bPathSync = <true>
2017-10-31 11:59:43.627 P00  DEBUG:     Storage::Local->new(): bAllowTemp = <true>, hRule = [undef], lBufferMax = 4194304, oDriver = [object], strDefaultFileMode = <0640>, strDefaultPathMode = <0750>, strPathBase = /db/9.4, strTempExtension = pgbackrest.tmp
2017-10-31 11:59:43.627 P00  DEBUG:     Storage::Local->openWrite(): bAtomic = <false>, bPathCreate = <false>, lTimestamp = [undef], rhyFilter = ({rxyParam => ({strCompressType => decompress}), strClass => pgBackRest::Storage::Filter::Gzip}), strGroup = [undef], strMode = <0640>, strUser = [undef], xFileExp = /var/lib/postgresql/TESTXLOG
2017-10-31 11:59:43.628 P00  DEBUG:     Storage::Base->copy(): xDestinationFile = [object], xSourceFile = [object]
2017-10-31 11:59:45.441 P00  DEBUG:     Archive::Get::Get->get=>: iResult = 0
```
The file copy takes just under two seconds.  Running with awscli I see the same result:

```
$ date +%H:%M:%S:%N;aws s3 cp s3://XXX/repo/archive/takeo/9.4-1/000000010000000B/000000010000000B00000054-b05a8fe51573d42b49e6bad71fd8ef44d05a9ee4.gz ~/TESTXLOG;date +%H:%M:%S:%N

11:56:16:765545089
download: s3://XXX/repo/archive/takeo/9.4-1/000000010000000B/000000010000000B00000054-b05a8fe51573d42b49e6bad71fd8ef44d05a9ee4.gz to ./TESTXLOG
11:56:18:759435122
```
Please run both of these commands on your system and attach the output.  Maybe you have a different bottleneck than the one I fixed.

**Comment by M1hacka** [comment]
pgbackrest:
```
2017-10-31 15:25:59.716 P00   INFO: archive-get command begin 1.25: --backup-host=db-backup --backup-user=ubuntu2 --db1-path=/vol1/postgresql/9.4/main --log-level-console=debug --log-level-file=trace --stanza=carrot
2017-10-31 15:25:59.757 P00   INFO: get WAL segment 00000005000038FD00000000
2017-10-31 15:25:59.757 P00  DEBUG:     Archive::Get::Get->get(): strDestinationFile = /var/lib/postgresql/TESTXLOG, strSourceArchive = 00000005000038FD00000000
2017-10-31 15:25:59.757 P00  DEBUG:     Protocol::Helper::protocolGet(): bCache = <true>, iProcessIdx = [undef], iRemoteIdx = <1>, strBackRestBin = [undef], strCommand = <archive-get>, strRemoteType = backup
2017-10-31 15:25:59.757 P00  DEBUG:     Protocol::Helper::protocolGet: create (cached) remote protocol
2017-10-31 15:25:59.758 P00  DEBUG:     Protocol::Remote::Master->new(): iBufferMax = 4194304, iCompressLevel = 6, iCompressLevelNetwork = 3, iProtocolTimeout = 1830, iSshPort = [undef], strCommand = /usr/bin/pgbackrest --buffer-size=4194304 --command=archive-get --compress-level=6 --compress-level-network=3 --db1-path=/vol1/postgresql/9.4/main --protocol-timeout=1830 --stanza=carrot --type=backup remote, strCommandSSH = ssh, strHost = db-backup, strUser = ubuntu2
2017-10-31 15:25:59.758 P00  DEBUG:     Protocol::Command::Master->new(): iBufferMax = 4194304, iCompressLevel = 6, iCompressLevelNetwork = 3, iProtocolTimeout = 1830, strCommand = ssh -o LogLevel=error -o Compression=no -o PasswordAuthentication=no ubuntu2@db-backup '/usr/bin/pgbackrest --buffer-size=4194304 --command=archive-get --compress-level=6 --compress-level-network=3 --db1-path=/vol1/postgresql/9.4/main --protocol-timeout=1830 --stanza=carrot --type=backup remote', strId = remote process on 'db-backup', strName = remote
2017-10-31 15:26:00.164 P00  DEBUG:     Protocol::Storage::Remote->new(): oProtocol = [object]
2017-10-31 15:26:00.164 P00  DEBUG:     Archive::Base->getCheck(): strDbVersion = [undef], strWalFile = 00000005000038FD00000000, ullDbSysId = [undef]
2017-10-31 15:26:00.164 P00  DEBUG:     Db::dbObjectGet(): bMasterOnly = true
2017-10-31 15:26:00.165 P00  DEBUG:     Db->new(): iRemoteIdx = 1
2017-10-31 15:26:00.165 P00  DEBUG:     Db::dbObjectGet=>: iDbMasterIdx = 1, iDbStandbyIdx = [undef], oDbMaster = [object], oDbStandby = [undef]
2017-10-31 15:26:00.165 P00  DEBUG:     Db->info(): strDbPath = </vol1/postgresql/9.4/main>
2017-10-31 15:26:00.165 P00  DEBUG:     Db->info=>: iDbCatalogVersion = 201409291, iDbControlVersion = 942, strDbVersion = 9.4, ullDbSysId = 6105204515923348092
2017-10-31 15:26:00.165 P00  DEBUG:     Protocol::Helper::protocolGet(): bCache = <true>, iProcessIdx = [undef], iRemoteIdx = <1>, strBackRestBin = [undef], strCommand = <archive-get>, strRemoteType = backup
2017-10-31 15:26:00.166 P00  DEBUG:     Protocol::Helper::protocolGet: found cached protocol
2017-10-31 15:26:00.684 P00  DEBUG:     Archive::Base->getCheck=>: strArchiveFile = 00000005000038FD00000000-0f33055f55cdd866401e5f6d9bd8b6620abe6f8d.gz, strArchiveId = 9.4-1
2017-10-31 15:26:00.684 P00  DEBUG:     Protocol::Storage::Remote->openRead(): rhParam = [hash], strFileExp = <REPO:ARCHIVE>/9.4-1/00000005000038FD00000000-0f33055f55cdd866401e5f6d9bd8b6620abe6f8d.gz
2017-10-31 15:26:01.119 P00  DEBUG:     Storage::Posix::Driver->new(): bFileSync = <true>, bPathSync = <true>
2017-10-31 15:26:01.119 P00  DEBUG:     Storage::Local->new(): bAllowTemp = <true>, hRule = [undef], lBufferMax = 4194304, oDriver = [object], strDefaultFileMode = <0640>, strDefaultPathMode = <0750>, strPathBase = /vol1/postgresql/9.4/main, strTempExtension = pgbackrest.tmp
2017-10-31 15:26:01.120 P00  DEBUG:     Storage::Local->openWrite(): bAtomic = <false>, bPathCreate = <false>, lTimestamp = [undef], rhyFilter = ({rxyParam => ({strCompressType => decompress}), strClass => pgBackRest::Storage::Filter::Gzip}), strGroup = [undef], strMode = <0640>, strUser = [undef], xFileExp = /var/lib/postgresql/TESTXLOG
2017-10-31 15:26:01.121 P00  DEBUG:     Storage::Base->copy(): xDestinationFile = [object], xSourceFile = [object]
2017-10-31 15:26:09.405 P00  DEBUG:     Archive::Get::Get->get=>: iResult = 0
2017-10-31 15:26:09.405 P00  DEBUG:     Common::Exit::exitSafe(): iExitCode = 0, oException = [undef], strSignal = [undef]
2017-10-31 15:26:09.405 P00  DEBUG:     Protocol::Helper::protocolDestroy(): bComplete = true, iRemoteIdx = [undef], strRemoteType = [undef]
2017-10-31 15:26:09.406 P00  DEBUG:     Protocol::Helper::protocolDestroy: found cached protocol: iRemoteIdx = 1, strRemoteType = backup
2017-10-31 15:26:09.414 P00  DEBUG:     Protocol::Command::Master->close=>: iExitStatus = 0
2017-10-31 15:26:09.414 P00  DEBUG:     Protocol::Helper::protocolDestroy=>: iExitStatus = 0
2017-10-31 15:26:09.414 P00  DEBUG:     Common::Lock::lockRelease(): bFailOnNoLock = false
2017-10-31 15:26:09.414 P00   INFO: archive-get command end: completed successfully
2017-10-31 15:26:09.414 P00  DEBUG:     Common::Exit::exitSafe=>: iExitCode = 0
```
aws download:
```
15:34:21:204896766
download: s3://cqbackup11/archive/carrot/9.4-1/00000005000038FD/00000005000038FD00000000-0f33055f55cdd866401e5f6d9bd8b6620abe6f8d.gz to ./TESTXLOG
15:34:27:847180516
```

**Comment by M1hacka** [comment]
Seems it's a network problem. I've repeated aws download several times and it took from 3 to 8 seconds...

**Comment by dwsteele** [comment]
OK, that's in line with what I'm seeing on my side.  If you are not running within AWS then S3 performance can be really spotty in my experience.

Parallelism  is generally the answer to this problem, but archive-get does not support parallel operation yet.  We are working on it and expect to release that feature this fall.  

Meanwhile, you might consider using more than one archive command to keep WAL local as well as pushing to S3.

**Comment by M1hacka** [comment]
I'll be greatfull, if you notify me on parallel archive-get released (m1ha@carrotquest.io).
Thanks.

**Comment by dwsteele** [comment]
Will do.  We are targeting parallel archive-get for the 2.0 release (January timeframe) but it might be a bit later.

**Comment by M1hacka** [comment]
Hi. 2.0 is released, but I don't see anything about parallel archive-get in release notes. Have anything changed?

**Comment by dwsteele** [comment]
Unfortunately, async archive-get did not make the cut for 2.0.  The good news is that it is at the top of the queue for 2.01 (or whatever the next feature release is).

There is an S3 performance improvement for 2.0 that may help archive-get some, but it may not be enough to make a real difference.

**Comment by dwsteele** [comment]
@M1hacka So, it took a while, but parallel archive-get was released in 2.02 and packages are available on apt.p.o and yum.p.o as of today.
